finetuner/tuner/pytorch/losses.py

```diff
@@ -174,3 +174,45 @@ def get_default_miner(
             return TripletMiner()
         else:
             return TripletSessionMiner()
+
+
+class NTXentLoss(nn.Module, BaseLoss[torch.Tensor]):
+    """Compute the NTXent (Normalized Temeprature Cross-Entropy) loss.
+
+    This loss function is a temperature-adjusted cross-entropy loss, as defined in the
+    `SimCLR paper <https://arxiv.org/abs/2002.05709>`. It operates on batches where
+    there are two views of each instance
+    """
+
+    def __init__(self, temperature: float = 0.5) -> None:
+        """Initialize the loss instance.
+
+        :param temerature: The temperature parameter
+        """
+        super().__init__()
+
+        self.temperature = temperature
+
+    def forward(self, embeddings: torch.Tensor, labels: torch.Tensor) -> torch.Tensor:
+        """Compute the loss.
+
+        :param embeddings: An ``[N, d]`` tensor of embeddings.
+        :param labels: An ``[N,]`` tensor of item labels. It is expected that each label
+            appears two times.
+        """
+        assert embeddings.shape[0] == labels.shape[0]
+
+        sim = (1 - get_distance(embeddings, 'cosine')) / self.temperature
+        diag = torch.eye(sim.shape[0], dtype=sim.dtype, device=sim.device)
+        labels1, labels2 = labels.unsqueeze(1), labels.unsqueeze(0)
+
+        pos_samples = (labels1 == labels2).to(sim.dtype) - diag
+
+        if not (pos_samples.sum(axis=1) == 1).all().item():
+            raise ValueError('There need to be two views of each label in the batch.')
+
+        self_mask = torch.ones_like(sim, requires_grad=False) - diag
+        upper = torch.sum(sim * pos_samples, dim=1)
+        lower = torch.log(torch.sum(self_mask * torch.exp(sim), dim=1))
```

Comment:
is it correct? I think `upper` should be `torch.exp(torch.sum(sim * pos_samples), dim=1)` and `lower` should be without `torch.log` and move it outside.

If I understand correctly, `upper` and `lower` means `numerator ` and `denominator` respectively